### PR TITLE
prevent static numpy array indices from being converted to tracers

### PR DESCRIPTION
**Pull request by jpbrodrick89**
I've been working on some function transformations to explore sparsity-leveraging accelerations in Jacobian computation and I've found without `EAGER_CONSTANT_FOLDING=1` static **numpy** array indices are being converted to tracers unnecessarily which prevents compile time inspection of sparsity (especially duplication). The conversion happens when indices are "normalised", e.g. their negative indices receive a positive shift equal to the length of array. Implementing this in pure numpy is trivial.

Let me know if I've missed any edge cases, but I think this should be a net win!

### CODE REVIEW

**Comment by jakevdp** [comment]
I don't think we should merge this: generally we treat NumPy arrays equivalently to JAX arrays to avoid potential surprises due to the implementation changing when an operation is wrapped in JIT. We should not start down the path of dispatching to NumPy when NumPy arrays are passed to JAX APIs.

**Comment by jpbrodrick89** [comment]
Thanks for the feedback @jakevdp, I'm aware that this is a bit more controversial than my other recent mini-PR's and the approach here might not be best. I appreciate that you want to ensure that  behaviour is the same whether you have a numpy or a Jax array. I think the main cognitive dissonance I'm struggling with is how this behaviour changes whether eager constant folding is on or not. If it's on then all constant index arrays are folded and we can do trace time inspection of them but without constant folding they're tracers.

Could you suggest any other more elegant ways to address. One option is to use pure Python primitives e.g. `(indices + size) % n` so that the overloaded operators are dispatched to automatically. If the onus has to be on the user that is fine too, but the main I would usually force selective constant folding is `with ensure_compile_time_eval`. But I can't index a dynamic array with a static one at compile time. 

Perhaps the user-owned way to address is just to call lax.gather directly ensuring all semantic demands are met, is that what you'd lean towards here? Although that would be against the recommendation in the docs "which is not something typical users should ever need to do" (although I'd accept I'm probably not a typical user 😅).

**Comment by jakevdp** [comment]
`(indices + size) % n` is semantically wrong: it means out-of-bound negative indices would be doubly-wrapped.

> just to call lax.gather directly

Note that the code this PR touches is used in `dynamic_slice`, not in `gather`.

You could ensure your static arguments are positive and pass `wrap_negative_indices=False` and skip this operation entirely – would that work?

**Comment by jpbrodrick89** [comment]
Thanks for the tip @jakevdp `.at[...].get(wrap_negative_indices=False)` did indeed skip this code path however, a more complicated problem is the `_index_to_gather` path which has two problematic paths.

1. `jax/_src/numpy/indexing.py:1313` — `lax_numpy.asarray(e.index)` unconditionally converts np→jax.
2. `jax/_src/numpy/indexing.py:1444` — `lax.expand_dims(g, (g.ndim,))` binds `broadcast_in_dim_p`, which under outer `jit` produces a tracer regardless of whether `g` was originally np.

2 can be trivially fixed by replacing the `expand_dims` with `g[..., None]` but I'm not sure what fixes you would be comfortable with  for 1 (an `isinstance(e.indx, np.ndarray)` skip path perhaps?)

Of course I can avoid all this by using `lax.gather` directly, but I'd rather follow the documentation and avoid it as far as possible :-) .

**Comment by jakevdp** [comment]
That makes sense – but I don't think we're going to change those code paths. Across JAX, we do no special handling of NumPy arrays, and instead convert them to JAX arrays at the first opportunity. This is to ensure that operations on those arrays are done on device (where they can be optimized by the compiler), rather than on host (where they lead to slow tracing and ofen turn into large compile-time constants that lead to poor performance).
